include/nsettings/settings.h

```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright (c) 2019 Laczen
+ * Copyright (c) 2018 Nordic Semiconductor ASA
+ * Copyright (c) 2015 Runtime Inc
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_SETTINGS_SETTINGS_H_
+#define ZEPHYR_INCLUDE_SETTINGS_SETTINGS_H_
+
+#include <sys/types.h>
+#include <misc/util.h>
+#include <misc/slist.h>
+#include <stdint.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @defgroup settings Settings subsystem
+ * @ingroup file_system_storage
+ * @{
+ */
+
+#define SETTINGS_MAX_DIR_DEPTH	8	/* max depth of settings tree */
+#define SETTINGS_MAX_NAME_LEN	(8 * SETTINGS_MAX_DIR_DEPTH)
+#define SETTINGS_MAX_VAL_LEN	256
+#define SETTINGS_NAME_SEPARATOR	"/"
+
+/* pleace for settings additions:
+ * up to 7 separators, '=', '\0'
+ */
+#define SETTINGS_EXTRA_LEN ((SETTINGS_MAX_DIR_DEPTH - 1) + 2)
+
+#define SETTINGS_NMGR_OP		0
+
+typedef ssize_t (*read_cb)(void *data, size_t len, void *cb_arg);
+
+/**
+ * @struct settings_handler
+ * Config handlers for subtree implement a set of handler functions.
+ * These are registered using a call to @ref settings_register.
+ */
+struct settings_handler {
+	sys_snode_t node;
+	/**< Linked list node info for module internal usage. */
+
+	char *name;
+	/**< Name of subtree. */
+
+	int (*h_get)(int argc, char **argv, void *val, int val_len_max);
+	/**< Get values handler of settings items identified by keyword names.
+	 *
+	 * Parameters:
+	 *  - argc - count of item in argv.
+	 *  - argv - array of pointers to keyword names.
+	 *  - val - buffer for a value.
+	 *  - val_len_max - size of that buffer.
+	 */
+
+	int (*h_set)(int argc, char **argv, size_t len, read_cb read,
```

Comment:
I don't understand what you mean. At the moment with the changes in the backend the h_set is not even called when the data in storage has been deleted.

Reply:
What if it was set before by FCB back-end?

Reply:
Good question. Do we want h_set() with len=0 to trigger a delete or not ? In the old settings it had to be done like this, in the new this is no longer the case. Do we want this functionality ? Extending this: how do we wan't to handle this with multiple sources ? Suppose the "factory setting" (first source) contains valid data while the second source (also the dst) contains deleted data, I would say it should then contain the "factory setting".

Reply:
I was not aware of that fcb back-end doesn't do that. Looks like we still need call h_set when using a "factory settings" (which is great idea IMHO). Problem is also implementation this for fcb - should be optional as it might increase application initialization a lot (I elaborate more on fcb-load implementation).